File: 20230516_Tuesday/20230516.rb
# Leetcode: 24. Swap Nodes in Pairs.
# https://leetcode.com/problems/swap-nodes-in-pairs/
# = = = = = = = = = = = = = =
# Accepted.
# Thanks God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 77 ms, faster than 86.27% of Ruby online submissions for Swap Nodes
# in Pairs.
# Memory Usage: 211 MB, less than 33.33% of Ruby online submissions for Swap
# Nodes in Pairs.
# 2023.05.16 Daily Challenge.
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} head
# @return {ListNode}
def swap_pairs(head)
  return head if ! head
  # Array to store pointers.
  arr = []
  # Save pointers to the array.
  tmp = head
  while tmp
    arr.push tmp
    tmp = tmp.next
  end
  # Nill all pointers in the array.
  arr.each{|x| x.next = nil}
  # Swap values in the each pair.
  i = 0
  while i < arr.size - 1
    arr[i],arr[i+1] = arr[i+1], arr[i]
    i += 2
  end
  # Redirect pointer.
  (0...arr.size-1).each do |i|
    arr[i].next = arr[i+1]
  end
  # Return a head.
  return arr[0]
end

File: 20230516_Tuesday/20230516.go
// Leetcode: 24. Swap Nodes in Pairs.
// https://leetcode.com/problems/swap-nodes-in-pairs/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 2 ms, faster than 62.89% of Go online submissions for Swap Nodes in
// Pairs.
// Memory Usage: 2 MB, less than 85.57% of Go online submissions for Swap Nodes
// in Pairs.
// 2023.05.16 Daily Challenge.
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func swapPairs(head *ListNode) *ListNode {
	// Do not work with empty list.
	if nil == head {
		return nil
	}
	// Prepare array for nodes.
	arr := make([]*ListNode, 101)
	sz := 0 // Size.
	// Fill the array.
	tmp := head
	for ; nil != tmp; sz += 1 {
		arr[sz] = tmp
		tmp = tmp.Next
	}
	// Nil all pointers in the nodes.
	// In sake of stability.
	for i := 0; i < sz; i += 1 {
		arr[i].Next = nil
	}
	// Swap things.
	for i := 0; i < sz-1; i += 2 {
		arr[i], arr[i+1] = arr[i+1], arr[i]
	}
	// Redirect pointers.
	for i := 0; i < sz-1; i += 1 {
		arr[i].Next = arr[i+1]
	}
	// Finaly return head.
	return arr[0]
}
